### ISSUE TITLE
optimizing numpy.random.choice() for user input probabilities.

### ISSUE THREAD

**Issue by Eight1911** [numpy/numpy#10432]
For random sampling from a user input distribution (when the keyword argument for `p` is not `None`), it seems that `numpy.random.choice` uses the bisection method, which runs in O(n log k) where  `n` is the number of samples and `k` the number of outcomes. 

We can do better. With O(k) preprocessing, the alias sampling method [(wiki)](https://en.wikipedia.org/wiki/Alias_method) has smaller constant and asymptotic cost even when k is small. (O(n) with at most two table lookups for each random sample.)

It would be more efficient to opt to use the alias method when sampling for a larger number of samples and the bisection method when the number of samples is small. If approved, I would like to work on this.

I can provide a C benchmark if needed.

**Comment by seberg** [comment]
This has been discussed before and is a good addition. The limiting factor is mostly the fact that you cannot just improve it! Before we can improve it we need the mechanisms to make sure that old random sequences will be stable (such as a protocol version, etc. with maybe some tricks that if you obviously do not care, we just use the faster version).
I think that has been worked on both for different generators as well as differen distribtions (effectively this is a distribution creation). Frankly not sure the status right now (it should be somewhere in the issues), just saying that you know that this is likely less a problem of improving the algorithm and more a problem of providing a stable API.

**Comment by seberg** [comment]
If you have the bandwidth, we are always happy about contributions!

**Comment by Eight1911** [comment]
After thinking 5 more minutes about it and rereading the [thread](http://numpy-discussion.10968.n7.nabble.com/Backwards-incompatible-improvements-to-numpy-random-RandomState-td40411.html), this might be easier than I had originally thought.

Without changing the RandomState API at all, we can add an attribute to RandomState (`_seeded` or something of the sort) to store whether a seed was explicitly set or given at initialization. Since we have no compatibility guarantee when a seed is not set, we can conditionally dispatch optimized or backward-compatible methods depending on the state of `self._seeded`. This idea is similar to what was discussed in the mailing list, but not exactly the same. But since there are no changes to the api, it won't be more complicated to adapt it later to fit the versioning method discussed in the mailing list.

This will allow the implementation of #2764 (floyd's algorithm for uniform sampling without replacement), and #2047 (ziggurat method for normal distribution) as well. I can get started right away if approved.

**Comment by seberg** [comment]
Probably true, though it would be nice if we find a solution that also works if you want to seed. After all, people should probably seed more rather then less ;). Anyway, frankly I do not know random that well or have a strong opinion, so not sure if it can be a good final solution.

**Comment by activatedgeek** [comment]
Hey @Eight1911, have you had a chance to rethink and make progress on this one?

**Comment by seberg** [comment]
There is a new NEP that will be put it soon, which describes how changing things in the random distributions will work (IIRC), so for anyone wanting to do something here, that would be a likely read.

**Comment by Eight1911** [comment]
@seberg The NEP mentioned the ongoing preparation is in `bashtage/randomgen` repo. Should I make a separate PR there?

**Comment by mattip** [comment]
@Eight1911 - yes

**Comment by mattip** [comment]
Closing as stale. Please restart the discussion in a new issue if still relevant.
